### ISSUE TITLE
[DISCUSSION] Migrating code out of sulfate_mod.F90 and gckpp_Hetrates.F90 into KPP

### ISSUE THREAD

**Issue by yantosca** [geoschem/geos-chem#567]
## Overview ##

We have set up this discussion thread to track progress of moving code out of sulfate_mod.F90 and gckpp_HetRates.F90 into KPP.  Please feel free to add to the discussion below in the comment boxes below.

## Original discussion thread ###

@tsherwen wrote:

>Please could I check if you are working on moving the sulfur chemistry from sulfate_mod into KPP? I took from a conversation with Viral Shah in Spring of this year that GCST were working on this. If this is the case, I'd be interested if you were able to give a quick status update on this?
> 
>I am interested right now as I am trying to use the simulation option with acidic uptake on dust. I generated an acid uptake run directory from the unit tester and ran this, but I see large decreases in NOy burdens globally (~15%) versus my base case (which should not be the happening). After digging into the code in sulfate_mod, I thought it made sense to check if these routines were being/had been moved into v12/KPP. I would also be interested in you know if you are aware of anyone using Ducan Farilie's code since it was merged into v11?
> 
>Related to this, I am aware that Linia Tashmim/Will Porter have also been working on moving and updated version of the sulfur chemistry [Chen et al. 2018] in KPP in v12. I assume if you are also working on sulfate_mod into KPP you are aware of their work too.

@yantosca replied:

>Most of the progress to date on this project has been in cleaning up the gckpp_Hetrates.F90.  We decided to not start the cleanup of gckpp_Hetrates until after the 12.9.0 release (Xuan Wang et al halogen chemistry).  As you know, the 12.9.0 release was delayed due to issues that came up during validation, and Will Downs spent a lot of time fixing those.
> 
>The problem was that soon after, we found that the new additions to the gckpp_Hetrates for 12.9.0 made the code  execute much more slowly than it should have.  I went in and made some fixes (basically pre-computing the indices and MWs of species instead of calling the Ind_() function all over the place) that made the code execute much faster.  That was a quick fix but I think I can do better.
> 
>Since then, the GCST has primarily been occupied with development for the upcoming 13.0.0 release.  Also note: I had to take about 2 months of sick leave off & on to support an ill family member.  So that has pushed the  schedule back even further.
> 
>I have started looking at the gckpp_Hetrates code and believe that I can pull most of the functions into the KPP rate law library (i.e. specified in #INLINE F90_RATES in gckpp.kpp).  But I am splitting time on this with other responsibilities.
> 
>We are also unware of other parallel efforts to bring the heterogeneous chemistry into KPP, but we would welcome assistance.  I personally do not have a very broad chemistry background, so I have been looking at this mostly from a coding perspective.
 
Viral Shah wrote:

>I know that Shixian Zhai has been using Duncan's dust acid uptake simulation in version 12.x

@zsx-GitHub wrote:

>This is Shixian Zhai. I have been using the acid uptake on dust option recently with GCv12.7.1. I'm not working on NOy so I don't know how it affects NOy. But I found that Ducan Farilie's original code generated too much sulfate on fine dust from the H2SO4 uptake on dust pathway. It turns out that this is because of its way of calculating the gas phase H2SO4 production rate in sulfate_mod.F. I fixed this by directly using the KPP H2SO4 production rate, since the gas phase SO2 oxidation is done in KPP. After my change, the model is simulating the on-dust acids reasonably. I am pasting what I changed in sulfate_mod.F for your reference. I also changed the way the model introduces dust alkalinity to be consistent with [Viral's paper](https://acp.copernicus.org/articles/20/12223/2020/acp-20-12223-2020.html). Hope this helps. 

```F90
         ! Isolate H2SO4 for reaction with dust    tdf 3/6/2K9
         IF ( LDSTUP ) THEN
!zhaisx 2020.10.16
!Now we change to directly using PSO4 [molec/cm3/s] from KPP
 
!            ! Compute gas phase SO4 production again, as in offline case
!            ! RK1: SO2 + OH(g) [s-1]  (rjp, bmy, 3/23/03)
             M    = State_Met%AIRDEN(I,J,L) * F
!            KK   = K0 * M / Ki
!            F1   = ( 1.e+0_fp + ( LOG10( KK ) )**2 )**( -1 )
!            RK1  = ( K0 * M / ( 1.e+0_fp + KK ) ) * 0.6e+0_fp**F1 *
!     &             GET_OH( I, J, L, Input_Opt, State_Chm, State_Met)
!            RKT  =  RK1 * DTCHEM  ! [unitless] (bmy, 6/1/00)
!            SO20 = SO2_cd
!            H2SO4_cd = SO20 * ( 1.e+0_fp - EXP( -RKT ) )
 
!Convert State_Diag%Prod from [molec/cm3/s] to [v/v/timestep].
            KPPProd0 = State_Diag%Prod(I, J, L, id_PSO4)
            H2SO4_cd = KPPProd0 / M * DTCHEM

            !tdf Reset these constants to zero to avoid any problems below
            M        = 0.e+0_fp
            KPPProd0 = 0.e+0_fp
!            KK  = 0.e+0_fp
!            F1  = 0.e+0_fp
!            RK1 = 0.e+0_fp
         ENDIF
```

@tsherwen wrote:

>Hello Bob, Shixian, and Viral,  
>
>Bob, Thanks a lot for your detailed response and especially the update on where things are with moving more of the mechanism into KPP. Hello Bob, Shixian, and Viral,  
>
>Bob, Thanks a lot for your detailed response and especially the update on where things are with moving more of the mechanism into KPP. The amount of work GCST manages to get through is always remarkable and the community is very appreciative. This year, more than ever, we must also put family first. I agree that creating a GitHub discussion issue for this seems like a good approach for linking up the various efforts related to this. 
>
>Viral and Shixan, thanks a lot for the response specifically about the acid uptake. Shixan, I’ve implemented the updates your shared and have a test run going on our HPC. I originally found large impacts using the acid-uptake directory from unit-tester (inc. on SO4), so will keep exploring this and it would be good to keep in contact with you for if you or I find any further issues.

**Comment by yantosca** [comment]
FYI, I am placing my updates in the *dev/bmy/epa-kpp* branch on Github.   You may see this branch at: https://github.com/geoschem/geos-chem/tree/dev/bmy/epa-kpp.**  

NOTE: I have included a fix for a parallelization error in sulfate_mod.F90 that had side-effects in the KPP solver.  See: https://github.com/geoschem/geos-chem/issues/392 for more information.

**Comment by wporter** [comment]
Hi Bob, thanks for pinging me on this topic!

Yes, as Tomas mentioned, my student Linia has been working on getting updated sulfur chemistry (specifically DMS oxidation) into FlexChem, and presented some of her initial progress at the recent GEOS-Chem Europe virtual meeting in September. So far she has added the previous updates of Chen et al., 2018 as well as the more recent HPMTF pathways from Veres et al., 2020, and performed some comparisons with previous implementations and observations. We'll be happy to share/compare/collaborate on this as needed!

**Comment by yantosca** [comment]
Thanks @wporter.  It would be great to see what Linia has done in the sulfate_mod.  I wasn't able to attend the GCE1 meeting virtually, so I might try to go back and look at her presentation on the Youtube site.

I have been working on cleaning up the `gckpp.kpp` and `gckpp_HetRates.F90` codes.  Right now I have noticed that there are a lot of unnecessary casts to from REAL\*4 to REAL\*8 (see the `DBLE()` calls) in the rate law library functions in `gckpp.kpp`. For example:

```F90
  ! Existing code

  REAL(kind=dp) FUNCTION ARRPLUS( A0,B0,C0,D0,E0 )
     REAL A0,B0,C0,D0,E0 
     REAL(kind=dp) K0     
     K0 = DBLE(A0) * (DBLE(D0)+(TEMP*DBLE(E0)))
     ARRPLUS =  K0 * EXP(-DBLE(B0)/TEMP) * (TEMP/300.0_dp)**DBLE(C0)
     IF (ARRPLUS < 0.0) ARRPLUS = 0.0
  END FUNCTION ARRPLUS        

 ! Call this with:
 0.063HO2 + 0.063CH2O + 0.937RIPA :          ARRPLUS(2.12E-13,-1300.0,0.0,1.1644,-7.0485E-4); {2019/11/06; Bates2019; KHB}
```
but as you can see, the many `DBLE()` casts can be eliminated by making sure to pass double-precision constant values such as:
```F90
  !  Code with DBLE() removed

  FUNCTION ARRPLUS( A0, B0, C0, D0, E0 ) RESULT( FVAL )
    REAL(kind=dp), INTENT(IN) :: A0,B0,C0,D0,E0
    REAL(kind=dp)             :: K0, FVAL
    K0   = A0 * (D0 + (TEMP*E0))
    FVAL = K0 * EXP(-B0/TEMP) * (TEMP/300.0_dp)**C0
    IF (FVAL < 0.0_dp) FVAL = 0.0_dp     
  END FUNCTION ARRPLUS

 ! Call this with
 0.063HO2 + 0.063CH2O + 0.937RIPA :          ARRPLUS(2.12E-13_dp, -1300.0_dp, 0.0_dp,  1.1644_dp, -7.0485E-4_dp); {2019/11/06; Bates2019; KHB}
```
where we use the `_dp`  suffix to indicate that the numerical constant should be treated as REAL*8 instead of REAL*4.  The `DBLE()` function takes a few CPU cycles to execute, and several `DBLE()` calls are done at each `(I,J.L)` grid box.  Removing these should result in a speedup (even a modest speedup would be great).

I also going to try to move all of the hetchem functions from `gckpp_HetRates.F90` into the `gckpp.kpp` file... this will place them into `gckpp_Rates.F90` when the mechanism is built.

Also note, if you are using the 12.9.0 code or later, I have fixed a pesky parallelization issue in `sulfate_mod.F90` that caused the same code to produce slightly different output (at the level of numerical noise).  See this issue: https://github.com/geoschem/geos-chem/issues/392.

**Comment by yantosca** [comment]
@kelvinhb: tagging you to this thread about moving code into KPP to keep you in the loop.

**Comment by yantosca** [comment]
Also note that the KPP code may not parse properly if you build it with a recent GNU or Intel compiler.  See: https://github.com/geoschem/KPP/issues/1

**Comment by yantosca** [comment]
I have also been trying to determine the impact of removing the DBLE() functions in the code.  I wrote a test function to compare how this is done in at present, and how it is done in my development code:

```F90
PROGRAM KppTest

  IMPLICIT NONE

  ! Parameters
  INTEGER, PARAMETER :: dp   = 8
  REAL*8,  PARAMETER :: TEMP = 293.0d0
  
  ! Variables
  REAL*8             :: ref, dev

  ref = GC1( 1.7E-11, 390.0,   9.33E-2, 5.05E15, -12200.0, 1.79E14, -8830.0  )
  dev = GC2( 1.7d-11, 390.0d0, 9.33d-2, 5.05d15, -1.22d4,  1.79d14, -8.830d3 )

  print*, 'Ref value: ', ref
  print*, 'Dev value: ', dev

CONTAINS

  REAL(kind=dp) FUNCTION GC1( A0,B0,C0,D0,E0,F0,G0 )
    ! GC_ISO1 from Ref
     REAL A0,B0,C0,D0,E0,F0,G0
     REAL(kind=dp) K0, K1, K2
     K0 = DBLE(D0)*EXP(DBLE(E0)/TEMP)*EXP(1.E8_dp/TEMP**3)
     K1 = DBLE(F0)*EXP(DBLE(G0)/TEMP)
     K2 = DBLE(C0)*K0/(K0+K1)
     GC1 =  DBLE(A0) * EXP(DBLE(B0)/TEMP) * (1.0_dp - K2)
  END FUNCTION GC1  

  FUNCTION GC2( a0, b0, c0, d0, e0, f0, g0 ) RESULT( rate )
    ! GC_ISO1 from dev
    ! Rate for ISOP + OH
    REAL(kind=dp), INTENT(IN) :: a0, b0, c0, d0, e0, f0, g0
    REAL(kind=dp)             :: k0, k1, k2, rate
    k0   = d0 * EXP( e0 / TEMP ) * EXP( 1.0E8_dp / TEMP**3 )
    k1   = f0 * EXP( g0 / TEMP )
    k2   = c0 * k0 / ( k0 + k1 )
    rate = a0 * EXP( b0 / TEMP ) * ( 1.0_dp - k2 )
  END FUNCTION GC2

END PROGRAM KppTest
```
Building and running this code gives
```
$ gfortran -O3 kpptest.F90 -o kpptest
$ ./kpptest 
 Ref value:    6.4255890436790718E-011
 Dev value:    6.4255889057583382E-011
```
As you can see, the Ref and Dev values are different.  This is due to the line:
```
  REAL A0,B0,C0,D0,E0,F0,G0
```
in the GC1 function.  Both the Intel and GNU Fortran compilers will interpret `REAL` as  `REAL*4` unless you tell it otherwise.  

To force `REAL` to be interpreted as `REAL*8`, we have to use these compile options:

 * Intel Fortran: `-r8`
 * GNU Fortran: `-fdefault-real-8 -fdefault-double-8`

If we use these options, we now get:
```
$gfortran -fdefault-real-8 -fdefault-double-8 kpptest.F90 -o kpptest
$ ./kpptest 
 Ref value:    6.4255889057583382E-011
 Dev value:    6.4255889057583382E-011
```
We should therefore add these lines to the `CMakeLists.txt` in the `KPP/fullchem` folder:
```CMake
target_compile_options(KPP_fullchem PRIVATE 
  ""
  # Treat REAL as if it were REAL*8
  $<$<STREQUAL:"${CMAKE_Fortran_COMPILER_ID}","Intel">:-r8>
  $<$<STREQUAL:"${CMAKE_Fortran_COMPILER_ID}","GNU">:-fdefault-real-8 -fdefault-double-8>
)
```
which will treat all REAL variables in the KPP-generated code as `REAL*8`.  I will also work on explicitly declaring all `REAL` variables with `REAL(kind=dp)` in the `gckpp.kpp` file.

**Comment by davenhenze** [comment]
Hi all! Some general thoughts about this effort. I believe that sulfate aq / het chem has been traditionally held separate from gas-phase chemistry in models like CMAQ, in order to prevent the gas-phase system from becoming overly stiff.  I'd be interested in to know if this is encountered in the current model dev here.  Perhaps GC's aq chem is not detailed enough to cause this problem? CMAQ's aq / het chem code is now entirely KPP based, but again I believe still separate from the gas-phase solver. See Fahey et al., Geosci Model Dev. 2017; 10(4): 1587–1605.  This includes non-equilibrium kinetic mass transfer, which I can imagine being included in GC eventually (e.g., Silver et al, ACP, 2017, doi:10.5194/acp-17-5107-2017).  While there are thus perhaps long-term performance concerns  for unifying gas / aq / het chem, kpp-ifying sulfate_mod.F90 and gckpp_Hetrates.F90 seems warranted regardless.

**Comment by yantosca** [comment]
Thanks @davenhenze.  For now I am just looking at structural code modifications.  There are some functions in gckpp_Hetrates that I think can go into the rate law library (whether the function is called in Hetrates or in gckpp_Rates shouldn't really matter).  There are other routines in sulfate_mod that we'll have to consider more closely, and whether we keep them separate from the gas-phase.  If I'm understanding correctly, maybe what we'd want to do is to have 2 KPP stacks, one for the gas-phase and one for the hetchem or sulfate_mod routines?

I haven't yet gotten to the point of looking how sulfate_mod routines can go into KPP.  I know that @wporter's group has been working on this.  We will connect w/ them after the winter break.

**Comment by davenhenze** [comment]
Sounds good Bob.  I guess I'm suggesting that eventually it might be desired to have 2 separate KPP stacks, or that you may find in the process of creating a single KPP mechanism for all chemistry it gets quite slow, even if it's been done correctly in terms of the code modifications.

**Comment by yantosca** [comment]
Thanks @davenhenze.  I think I see see what you mean now.  In GeosCore/flexchem_mod.F90 we call SET_HET (in gckpp_HetRates.F90) to update the HET array, and then UPDATE_RCONST to copy the het rates into the RCONST array.  That was presumably done to avoid calling functions again deep down into KPP.

But... there is a section later down that, in case KPP can't converge to a solution, it calls UPDATE_RCONST one more time (but not SET_HET) to recompute rate constants.  If it can't converge after the 2nd pass,

So if we bring the functions in SET_HET into gckpp_Rates, then we might incur a penalty of having to compute them twice at most for each grid box.  This might not be too bad.  They say that "flops are free but memory costs", i.e. it can be potentially cheaper to recompute something twice in parallel code than to have to read it from memory.  We can see how it goes.

**Comment by yantosca** [comment]
Also -- chemistry question.  I notice that we have the Arrhenius function GCARR defined in gckpp.kpp, which takes the usual 3 Arrhenius coefficients (A, B, C):

https://github.com/geoschem/geos-chem/blob/681baa91d5e72c54c059acee22acc4bc48deebd2/KPP/fullchem/gckpp.kpp#L125-L129

But it seems that in the current fullchem.eqn file in the mechanism that we use, the Arrhenius B coefficient is always zero:

https://github.com/geoschem/geos-chem/blob/681baa91d5e72c54c059acee22acc4bc48deebd2/KPP/fullchem/fullchem.eqn#L433-L1309

If that is the case, I would propose commenting out changing the function to:
```F90
FUNCTION GCARR( a0, c0 ) RESULT( rate ) 
   ! Arrhenius function 
   REAL(kind=dp), INTENT(IN) :: a0, c0
   REAL(kind=dp) :: rate
   rate =  a0 * EXP ( c0 / TEMP ) 
 END FUNCTION GCARR 
```
(or even write a second parallel function).  Then the function calls can be like:
```
 O3 + NO = NO2 + O2 :                         GCARR(3.00d-12,  -1500.0d0); 
```
so that we skip passing the B argument altogether.  I'm just thinking that this would be more efficient than always having to do that division and exponentiation by 0, which results to 1 anyway.  Better to avoid useless CPU cycles that get done on each (I,J,L) iteration.

Thoughts?

**Comment by wporter** [comment]
That change makes sense to me, Bob. If the removed term were ever needed in the future for some reason, the expanded version could always be added back in for the relevant reactions, right? Might as well stop multiplying by one where possible to improve computational efficiency and mechanism clarity.

**Comment by yantosca** [comment]
Thanks @wporter.  I'm actually keeping the GCARR function as it is and making a parallel GCARR2 function.  So the user still use GCARR if all 3 coefficients are presnet, or GCARR2 if only A and C are present.

Another modification:  if both Arrhenius B and C coefficients are zero, we can just skip the function call and just assign the rate constant as the A coefficient, i.e.:

```
 HCOOH + OH = H2O + CO2 + HO2 :               GCARR(4.00E-13, 0.0E+00, 0.0);       {2013/03/22; Paulot2009; FP,EAM,JMAO,MJE} 
```
would become
```
 HCOOH + OH = H2O + CO2 + HO2 :               4.00E-13;       {2013/03/22; Paulot2009; FP,EAM,JMAO,MJE} 
```
because the EXP( C / T ) term goes to 1 in this case.

**Comment by wporter** [comment]
Ah yes, that seems like an ideal solution. Use simplified function calls where possible, keep the expanded versions for use where necessary.

Any sense as to how much impact these changes would have in practice, though? I'm always curious. :)

**Comment by yantosca** [comment]
Thanks @wporter.  I sense they could have a big impact, any little savings will be multiplied as we go to finer & finer resolution.  I can do a profiling of the code before & after to quantify it.

**Comment by tsherwen** [comment]
This sounds like a really good option @yantosca. I too would be interested in how much of a difference this makes. Leaving the existing function present and just adding a simpler function to use for rates with fewer terms sounds good. It might also be a good opportunity for us to add some more documentation to the gckpp.kpp file as that might make it more accessible for people to edit. 

Do you know whether this specific change was highlighted by Mike Long/John Linford's work on reducing the KPP files complexity before running the calculation with KPPA? Moreover, are there any other updates we could make that were learnt from the KPPA work?

**Comment by davenhenze** [comment]
@yantosca -- with regards to your earlier question, I was thinking about the efficiency of the overall chemical owing to increasing the stiffness of the ODEs when adding cloud chemistry. This could cause more / shorter internal time steps or a need for smaller ATOL and RTOL values for stability. 

However, I agree with your suggestions for making Update_RCONST more efficient, and the support from others here. 

One additional note pertaining to your earlier comment: in the case that IERR < 0 (i.e. failed ODE solve), I don't think Update_RCONST needs to be called again.  The values of RCONST should not be affected by any calls to the integrator, because RCONST are not updated at internal time steps within the integrator.  I originally reset RCONST under these conditions when I first implemented this, cautiously, but I suspect it isn't strictly necessary.

**Comment by yantosca** [comment]
Hi @wporter.  Just wanted to touch base now that we're back from the holidays.  We would be greatly interested in collaborating with you on the efforts to bring sulfate_mod code into KPP.

Also, I have been able to remove most of the unnecessary computations in KPP rate law routines by adding new functions that skip computing terms that would evaluate to 1.  I still need to do timing/profiling tests but I think it should have a speedup.

**Comment by wporter** [comment]
Hi @yantosca,

Glad to hear about the computation improvements -- will be cool to see the timing impacts!

I had a quick (and hopefully not TOO stupid) question to ask before starting into the KPP change details. In sulfate_mod, the CHEM_DMS subroutine currently appears to only get called during offline aerosol simulations, which would make sense because the 3 reactions it includes (abstraction/addition for DMS + OH, along with DMS + NO3) are already present in standard chemistry KPP (though the yields and constants appear to have been updated in the KPP version). Am I interpreting this correctly? If so, what portions of DMS chemistry in sulfate_mod are you hoping to migrate over to KPP -- is this still for aerosol-only simulations?

Linia's work has focused on building off of the existing three DMS reactions that are already in KPP to include the updates from Chen et al., 2018 (previously added to v9-02), as well as the more recent chemistry from Veres et al., 2020. We haven't touched any of the acid uptake portions of sulfate_mod, nor the DMS chemistry that's still in there for aerosol-only simulations, but we'll be happy to help test the updates suggested in this thread, and to contribute towards any other changes that would be useful.

**Comment by yantosca** [comment]
Thanks @wporter.  You are right, the CHEM_DMS subroutine is called for the aerosol-only module.  It may be a bit of historical baggage, I think CHEM_DMS might have also been used in the fullchem simulation way back when, but then those 3 reactions were added into the SMVGEAR mechanism (and now the KPP mechanism),  But we left it for the aerosol-only.  

Ideally we would like to move as much as is possible out of sulfate_mod and into the KPP mechanism (e.g. CHEM_SO2, CHEM_SO4, CHEM_MSA, etc.)  I understand that there might be some subtleties given that some of these reactions might also cause the chemical mechanism to get too stiff.  As I have a limited understanding of chemistry, I don't have a good sense of where the potential roadblocks would be.  Maybe this is where we can leverage the experience that you & Linia have from her past work.

This is also with a view of moving the specialty simulation code into KPP mechanisms.  Ideally the aerosol-only would be a subset of the full-chemistry (to the greatest extent possible).  The CO2, CO, and CH4 mechanisms could likewise be folded into a single mechanism that covers all those species.  That's the idea, anyway.

My updates are in the dev/epa-kpp branch of the GEOS-Chem and GCClassic repositories.  I am still testing it and trying to nail down a numerical drift issue.  I'll continue to post updates here.

**Comment by wporter** [comment]
Ahh, got it -- that makes sense, and we can certainly help with the shift and testing of the chemistry from those subroutines. Do you have a preferred order of operations as far as priorities for the migration? CHEM_SO2 is going to be a real beast, but others like CHEM_MSA seem more or less ready to disappear. Also, although I've heard about the goal of transitioning specialty simulations as chemistry "subsets", I'm not up-to-date on the logistics of how to best implement this in practice. Is there a set of slides somewhere that outlines the preferred way for us to define these kinds of special cases within KPP, moving forward?

**Comment by yantosca** [comment]
Thanks @wporter.  I think I am going to try to tackle gckpp_HetRates.F90 and sulfate_mod.F90 first, at least with the low-hanging fruit.  CHEM_MSA and similar routines sound like a good place to start.

For the specialty simulations I don't have slides but I guess the idea would be to try to isolate the chemical reactions in those routines that are explicitly written out and add them as KPP equations (with corresponding rate law functions).

Also, I should have profiling results soon.

**Comment by yantosca** [comment]
I have posted some profiling results that assesses the performance of removing the useless computations at feature request #598.

**Comment by yantosca** [comment]
Tagging @viral211 on this thread.

**Comment by yantosca** [comment]
As it turns out, the timing results reported in #598 were not an apples-to-apples comparison. The KPP integrator routine in the Dev code was calling UPDATE_RCONST (i.e. the rate law update function) within the integrator on each integration step. Although the time to compute the reaction rates improved, the overall integration time increased.

A new test has been done that shows removing the useless computations in the rate law functions actually decrease the time by about 4.5 minutes per 7 day simulation (or almost 20 minutes for a 31-day benchmark simualtion). See this link: #598 (comment)
